validator/client/sync_committee.go

```diff
@@ -0,0 +1,225 @@
+		return
+	}
+	sszRoot := types.SSZBytes(res.Root)
+	r, err := helpers.ComputeSigningRoot(&sszRoot, d.SignatureDomain)
+	if err != nil {
+		log.WithError(err).Error("Could not get sync committee message signing root")
+		return
+	}
+	sig, err := v.keyManager.Sign(ctx, &ethpb.SignRequest{
+		PublicKey:       pubKey[:],
+		SigningRoot:     r[:],
+		SignatureDomain: d.SignatureDomain,
+	})
+	if err != nil {
+		log.WithError(err).Error("Could not sign sync committee message")
+		return
+	}
+
+	msg := &ethpb.SyncCommitteeMessage{
+		Slot:           slot,
+		BlockRoot:      res.Root,
+		ValidatorIndex: duty.ValidatorIndex,
+		Signature:      sig.Marshal(),
+	}
+	if _, err := v.validatorClient.SubmitSyncMessage(ctx, msg); err != nil {
+		log.WithError(err).Error("Could not submit sync committee message")
+		return
+	}
+
+	log.WithFields(logrus.Fields{
+		"slot":           msg.Slot,
+		"blockRoot":      fmt.Sprintf("%#x", bytesutil.Trunc(msg.BlockRoot)),
+		"validatorIndex": msg.ValidatorIndex,
+	}).Info("Submitted new sync message")
+}
+
+// SubmitSignedContributionAndProof submits the signed sync committee contribution and proof to the beacon chain.
+func (v *validator) SubmitSignedContributionAndProof(ctx context.Context, slot types.Slot, pubKey [48]byte) {
+	ctx, span := trace.StartSpan(ctx, "validator.SubmitSignedContributionAndProof")
+	defer span.End()
+	span.AddAttributes(trace.StringAttribute("validator", fmt.Sprintf("%#x", pubKey)))
+
+	duty, err := v.duty(pubKey)
+	if err != nil {
+		log.Errorf("Could not fetch validator assignment: %v", err)
+		return
+	}
+
+	indexRes, err := v.validatorClient.GetSyncSubcommitteeIndex(ctx, &ethpb.SyncSubcommitteeIndexRequest{
+		PublicKey: pubKey[:],
+		Slot:      slot,
+	})
+	if err != nil {
+		log.Errorf("Could not get sync subcommittee index: %v", err)
+		return
+	}
+	if len(indexRes.Indices) == 0 {
+		log.Debug("Empty subcommittee index list, do nothing")
+		return
+	}
+
+	selectionProofs := make([][]byte, len(indexRes.Indices))
+	for i, index := range indexRes.Indices {
+		subCommitteeSize := params.BeaconConfig().SyncCommitteeSize / params.BeaconConfig().SyncCommitteeSubnetCount
+		subnet := uint64(index) / subCommitteeSize
+		selectionProof, err := v.signSyncSelectionData(ctx, pubKey, subnet, slot)
+		if err != nil {
+			log.Errorf("Could not sign selection data: %v", err)
+			return
+		}
+		selectionProofs[i] = selectionProof
+	}
+
+	v.waitToSlotTwoThirds(ctx, slot)
+
+	for i, comIdx := range indexRes.Indices {
+		isAggregator, err := altair.IsSyncCommitteeAggregator(selectionProofs[i])
+		if err != nil {
+			log.Errorf("Could check in aggregator: %v", err)
+			return
+		}
+		if !isAggregator {
+			continue
+		}
+		subCommitteeSize := params.BeaconConfig().SyncCommitteeSize / params.BeaconConfig().SyncCommitteeSubnetCount
+		subnet := uint64(comIdx) / subCommitteeSize
+		contribution, err := v.validatorClient.GetSyncCommitteeContribution(ctx, &ethpb.SyncCommitteeContributionRequest{
```

Comment:
How many requests will we be making on average here? Maybe in the future could involved some batching

Reply:
This is a very rare event, I rather keep it simple.

64 out of N on per slot base. N is the total validator count